### ISSUE TITLE
Improve drive start recording

### ISSUE THREAD

**Issue by marcogabriel** [adriankumpf/teslamate#425]
Teslamate polls frequently if the car is awake or sleeps, before requesting data that would wake up the car. If a drive is started, Teslamate possibly misses the first few minutes of the drive. I know, there are some solutions for this problem like installing Tasker on Android to inform Teslamate immediately when the mobile phone connects to the Tesla Bluetooth connection and some other ways to do this.

While this works, it's somehow cumbersome to set up and depends on the users experience.

Maybe it could be changed to work like this:

* teslamate frequently polls the status of the car
* status == (parked and asleep) -> do nothing, wait for next check
* status change from (parked and asleep) to (parked and awake) -> take current position as starting point of next drive - this would still be possible if using Tasker or something similar
* *NEW* status change from (asleep and parked) to (awake and driving) -> take endpoint of last drive as starting point of current drive

Advantages:

* this would Improve the accuracy of drives
* you can still use something like Tasker etc. if there is a reason for it
* you don't have to use something like Tasker because the car already knows where it was parked by the end of the previous drive

Disadvantages:

* if the car gets moved without starting it like it gets hauled off, the starting point of the next drive may be false - but this is most probably uncommon

EDIT:
If this can/will be implemented: A migration of existing data would be very nice. AFAIK, the necessary data for former drives is already in the database (shown in states dashboard). I am not sure if this data is sufficient for a migration of starting points.

EDIT 2:
As an additional factor, the mileage can help to decide if the drive was started earlier. If end mileage of last drive < start mileage of current drive -> set starting point to endpoint of last drive.

**Comment by olexs** [comment]
The issue here is, Teslamate _must_ pause the polling in order to allow the car to sleep. That wait period ("falling asleep") is the only time when a drive start can be missed - once the period is over and the API returns "asleep", Teslamate polls every 30 seconds. If Teslamate doesn't pause polling for the "falling asleep" period, the car will never sleep - this is an issue with the Tesla API, it will keep the car awake and only allow it to fall asleep if no API requests are made for a period of time.

**Comment by marcogabriel** [comment]
Thanks for clarification. 

From my recorded data, my starting point nearly every morning isn't my defined geofenced "Home" but a position that's not far away. Maybe one or two kilometers away.

This behaviour could be addressed and improved by the change above.

**Comment by ngardiner** [comment]
I don't believe that Teslamate does stop polling the vehicle state for that
period, only the more intensive data logging period. This is based on a
response that @adriankumpf gave recently to a suggestion about time-based
wake ups where he explained this is not necessary as vehicle state is
constantly polled from an API endpoint that does not wake the vehicle even
when logging is suspended, and logging resumes as soon as it sees the state
change. Perhaps @adriankumpf could confirm how that would align here, as
you would think that logic would be sufficient to have detected the vehicle
being prepared for the drive within 1 minute of unlocking/waking it?

**Comment by olexs** [comment]
> From my recorded data, my starting point nearly every morning isn't my defined geofenced "Home" but a position that's not far away. Maybe one or two kilometers away.

This happens often to me as well. The reason is, at least for me, not Teslamate but the car itself: while it sits asleep/offline in the garage, it loses GPS reception, and only regains a correct position several minutes into a "cold start" drive. This can be seen on the in-car nav screen - it keeps showing the car in a completely wrong place, until it eventually "jumps" to the correct location once the GPS reception has picked back up. I believe that's the point when Teslamate gets the first valid position update, and the drive starts recording.

Your proposed solution of taking the last known position before the car went asleep/offline instead of the first position of the new drive could actually  solve this, at least in part.

**Comment by marcogabriel** [comment]
added two additional edits:

EDIT:
If this can/will be implemented: A migration of existing data would be very nice. AFAIK, the necessary data for former drives is already in the database (shown in states dashboard). I am not sure if this data is sufficient for a migration of starting points.

EDIT 2:
As an additional factor, the mileage can help to decide if the drive was started earlier. If end mileage of last drive < start mileage of current drive -> set starting point to endpoint of last drive.

**Comment by codingoverdrive** [comment]
I recently installed Teslamate and I think it's great.

I have also found that Teslamate is definitely missing the start of drives even when the car has a GPS lock - so this is not the cause of the issue (as reported by @olexs above). 

My understanding is that `https://owner-api.teslamotors.com/api/1/vehicles/` can be called frequently and without affecting the awake/sleep status of the car. The status will be one of: `online, offline, asleep, waking, unknown`

Could it be the case that Teslamate is not calling the API above frequently enough to correctly identify that the vehicle has woken up and/or is about to start?

Once it detects a change in state, it should start the sleep timer and poll frequently for a change in drive state using `https://owner-api.teslamotors.com/api/1/vehicles/{{vehicle}}/data_request/drive_state` 

If there is no change in the drive state after the sleep window timeout, then it should revert to the `../vehicles/` API call to avoid preventing the vehicle from sleeping again.

I believe that the behaviour of vehicles and sleep mode varies from MCU1 (Model S/X) to the newer MCU (>Apr 2018 S/X and M3)
